Core/clim-basic/utils.lisp

```diff
@@ -468,10 +468,27 @@ in KEYWORDS removed."
 
 ;;;; ----------------------------------------------------------------------
 
+;;; FIXME valid space specification format is described in the section
+;;; describing a FORMATTING-TABLE macro. This should be generalized
+;;; for other possible space definitions and described in a separate
+;;; section. This functionality partially overlaps with a space
+;;; specification format described for the layout macros like
+;;; VERTICALLY, We should scram a superset of both in PARSE-SPACE and
+;;; add a special handling for the non-stream panes. -- jd 2019-11-02
+
+(deftype space-spec ()
+  `(or real
+       string
+       character
+       function
+       (cons real
+             (cons (member :character :line :point :pixel :mm)
+                   null))))
```

Comment:
Generators for testing `parse-space' directly or for improving the `normalize-margin-spec` test:

```lisp
(defun gen-unit ()
  (gen-one-element :character :line :point :pixel :mm))

(defun gen-invalid-space-spec ()
  (let ((value (gen-integer :min 0 :max 100))
        (unit (gen-unit)))
    (lambda ()
      (case (random 5)
        (0 :invalid-space)
        (1 #C(1 2))
        (2 (list (funcall value) :invalid-unit))
        (3 (list :invalid-real (funcall unit)))
        (4 (list (funcall value) (funcall unit) 3))))))

(defun gen-invalid-margin-spec-element ()
  (let ((anchor (gen-one-element :relative :absolute))
        (space (gen-invalid-space-spec)))
    (lambda ()
      (case (random 3)
        (0 :foo)
        (1 (list (funcall anchor) 1 2))
        (2 (list (funcall anchor) (funcall space)))))))
```

Reply:
added to utils.lisp test